api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java

```diff
@@ -1153,8 +1168,166 @@ public Appointment bookAppointment(Appointment appointment,
+		return getAppointmentResourceDAO().getAll();
+	}
+
+	@Override
+	public List<AppointmentResource> getAllAppointmentResources(boolean includeVoided) {
+		return getAppointmentResourceDAO().getAllData(includeVoided);
+	}
+
+	@Override
+	public AppointmentResource getAppointmentResource(Integer appointmentResourceId) {
+		return (AppointmentResource) getAppointmentResourceDAO().getById(appointmentResourceId);
+	}
+
+	@Override
+	public AppointmentResource getAppointmentResourceByUuid(String uuid) {
+		return (AppointmentResource) getAppointmentResourceDAO().getByUuid(uuid);
+	}
+
+	@Override
+	public AppointmentResource saveAppointmentResource(AppointmentResource appointmentResource) throws APIException {
+		appointmentResource.setIncludeWeekends(true);
+		ValidateUtil.validate(appointmentResource);
+		return saveAppointResourceInternal(appointmentResource);
+	}
+
+	@Override
+	public AppointmentResource voidAppointmentResource(AppointmentResource appointmentResource, String reason) {
+		appointmentResource.setVoided(true);
+		appointmentResource.setVoidReason(reason);
+		return (AppointmentResource) getAppointmentResourceDAO().saveOrUpdate(appointmentResource);
 	}
 
+	@Override
+	public void purgeAppointmentResource(AppointmentResource appointmentResource) {
+		getAppointmentResourceDAO().delete(appointmentResource);
+	}
+
+	@Override
+	public List<AppointmentResource> getAppointmentResourcesByConstraints(Location location, Provider provider, List<AppointmentType> appointmentTypes) {
+		return getAppointmentResourceDAO().getResourceByConstraints(null, null, null, null, location, null);
+	}
+
+	@Override
+	public BlockExcludedDays geyExcludedDayByUuid(String uuid) throws APIException {
+		return appointmentResourceDAO.geyExcludedDayByUuid(uuid);
+	}
+
+	private synchronized AppointmentResource saveAppointResourceInternal(AppointmentResource appointmentResource) throws APIException {
+		//check if resource already exists
+		List<AppointmentResource> resource = getAppointmentResourceDAO().getResourceByConstraints(appointmentResource.getStartDate(), appointmentResource.getEndDate(),
+				appointmentResource.getStartTime(), appointmentResource.getEndTime(), appointmentResource.getLocation(), null);
+		if (resource.size() > 0) {
+			throw new APIException("Resource with those details already exist");
+		} else {
+			if (CollectionUtils.isEmpty(appointmentResource.getDaysList())) {
+				return (AppointmentResource) getAppointmentResourceDAO().saveOrUpdate(appointmentResource);
+			}
+			appointmentResource.setDaysList(null);
+
+			return (AppointmentResource) getAppointmentResourceDAO().saveOrUpdate(appointmentResource);
+		}
+
+	}
+
+
+	public TimeSlot createTimeSlot(Date date, Location location) {
+
+		List<AppointmentResource> ar = getAppointmentResourceDAO().getResourceByConstraints(null, null, null, null, location, date);
+		if (ar != null) {
+			TimeSlot timeSlot = new TimeSlot();
+			for (AppointmentResource appointmentResource : ar) {
```

Comment:
I'm a little confused, looks like you loop through all the possible resources here, and within the loop you create a appointment block, so you'd be creating a block-per-resource, but then the Time Slot is only saved when you are outside the loop?  So the time slot would only be for the last appointment block.

I would think you would just pick the "most relevant" appointment resource and create a single block and time slot?  (We'd need to determine how the "most relevant" would be determined)